### PR TITLE
Section on total functions + smaller changes

### PR DESCRIPTION
**Pull request by ybertot**
re-work the explanation of data types so that the analogy with the axioms
of Peano arithmetic is used more efficiently.

Adds a paragraph before addressing the data types of boolean values and
natural numbers, so that later paragraphs that introduce language constructs
without respecting the separation between bool and nat feel less awkward.

Adds an explanation of the .+2 notation.

Adds a section about total functions (1 page and a half).

### CODE REVIEW

File: tex/chProgramming.tex [hunk 1]
```diff
@@ -618,6 +620,11 @@ \section{Data types, first examples}
 sequences of values are among the most prominent examples.
 \index[concept]{inductive type}
 
+In the following sections, we introduce basic data types \C{bool} for boolean
+values and \C{nat} for natural numbers, taking the opportunity to describe
+various constructs of the \Coq{} language as the become relevant.  These
```
Comment (amahboubi) [inline tex/chProgramming.tex]: as theY become?

File: tex/chProgramming.tex [hunk 2]
```diff
@@ -618,6 +620,11 @@ \section{Data types, first examples}
...
+sections serve simultaneously as an introduction to the data types and to the
```
Comment (amahboubi) [inline tex/chProgramming.tex]: I do not understand "to the data types and to the Coq language constructs". Does this mean something like "to the definition of inductive data types in Coq"?

File: tex/chProgramming.tex [hunk 3]
```diff
@@ -768,7 +776,7 @@ \subsection{Natural numbers}\index[concept]{natural number}
 
 When defining functions that operate on natural numbers, we can
 proceed by case analysis, as was done in the previous section for boolean
-values. Here as well, there are two cases: either the natural number used in
```
Comment (amahboubi) [inline tex/chProgramming.tex]: I thought that  "here as well' was more idiomatic than "here also", but I trust your English better than mine :)

File: tex/chProgramming.tex [hunk 4]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+
+Usual mathematical practice, learned by
+everybody from middle school, is that we can consider a bigger type of
+numbers, in which the natural numbers are injected implicitly, and
+where the predecessor of \C{0} and the results of exceptional
+subtractions do exist.  So the predecessor of 3 is a natural number,
+but the predecessor of \C{0} is not, it is a relative integer.  This
+versatility, allowed by mathematical practice, may seem to provide
+the necessary soupleness to be able to write mathematical formulas
+quickly and meaningfully.  It is not the case.  there are mathematical
```
Comment (amahboubi) [inline tex/chProgramming.tex]: Missing capital here: There are...

File: tex/chProgramming.tex [hunk 5]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+functions that are inherently not defined everywhere, in ways that
+cannot be recovered by just adding a new type.
```
Comment (amahboubi) [inline tex/chProgramming.tex]: But this does not look like an objection to the fact that mathematical formulas written (quickly :)) on paper are meaningful. Does it?

Comment (darijgr) [inline tex/chProgramming.tex]: Lines 1282-1290: Does this really mean that defining a ring in Coq requires specifying an inverse function? What about rings where it is not known in advance which elements are invertible and which are not? Say, some quotient of a multivariate polynomial ring...

File: tex/chProgramming.tex [hunk 6]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+
+The example we will choose is the example of multiplicative inverse.
+For natural numbers, only \(1\) has a multiplicative inverse
+\(1^{-1}=1\), so that for every number \(n\), \(n \times 1 \times
+1^{-1} = n\).  In a suitably large numerical type, such as a type of
```
Comment (anton-trunov) [inline tex/chProgramming.tex]: `In a suitably large numerical type...`
But the cardinality of the set of natural numbers is the same as the cardinality of the rationals.
Maybe I misinterpreted this part.

File: tex/chProgramming.tex [hunk 7]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+rational numbers, almost any number \(a\) has a multiplicative inverse
+\(a^{-1}\) so that for every other number \(n\) \(n \times a \times
+a^{-1}= n\).  It is only needed that \(a\) is non zero.  But zero
+remains an ``exceptional'' case, whatever enlargement of the output
+type we may propose.
+
+In the case of the inverse of zero, the usual mathematical practice
+takes a different stance.  This formula is just not defined, and
+similarly for any number \(n\), \(n/0\) is not defined.  In this
+sense, usual mathematics prefer to take a risk.  Mathematicians are
```
Comment (anton-trunov) [inline tex/chProgramming.tex]: A typo: preferS

File: tex/chProgramming.tex [hunk 8]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+allowed to write formula using division, but when it comes to
```
Comment (anton-trunov) [inline tex/chProgramming.tex]: typo: formulaS

File: tex/chProgramming.tex [hunk 9]
```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
...
+understanding the meaning of these formulas, it is their
+responsibility to check that the formula makes sense, in other words,
+that the formula does not contain any division by zero.  Thus a given
+formula is not well defined in all cases, it is well defined in
+precise context, the one that guarantees that there is
+no division by zero.
+
+In the \Coq{} system, the approach is different: every mathematical
+formula must have a well-defined value.  When facing a partially
+defined function, this means that a neigbour total function is chosen,
```
Comment (anton-trunov) [inline tex/chProgramming.tex]: neigHbour

File: tex/chProgramming.tex (hunk 5 above)
Comment (ybertot) [inline tex/chProgramming.tex]: There are several ring structures in the Math-comp library.  The "unit ring structure" is one where one explicitly decide to have a multiplicative inverse function and a predicate "unit" which describes precisely those elements for which the result of the function is meaningful.  Not all rings need to implement the "unit ring structure", but having this structure independently from the Field structure is useful.

File: tex/chProgramming.tex (hunk 3 above)
Comment (darijgr) [inline tex/chProgramming.tex]: Or "Here again"?

**Comment by ybertot** [comment]
Actually, the word language was used from line 587, and then repeated on 591, 595.  I could use the word "theories" instead, but I am not sure this will bring less confusion.  I propose we stay with the "languages" word.

I am more concerned by your remark about "rules on how to compute".  You are right that there should not be any relation between constructors (the data-type is very close to a free algebra).  However, each data-type is equiped with a generic mechanism to define functions by case analysis and recursion, and this is what I wanted to refer to.  We need to find a better phrasing.  Thanks for pointing that out.

**Comment by ybertot** [comment]
After discussing with Assia, we concluded that the part on total function should be moved to another chapter (Chapter 3).

**Comment by darijgr** [comment]
Ah, that's what you mean! Then I would just say it like that: "rules for defining functions from this data type".

Anyway, it probably does make sense to move this to a later chapter, since it's less important for understanding basic proofs. Let me know when it's closer to being done (and no longer conflicts with master). Thanks!

**Comment by ybertot** [comment]
This pull request should now be canceled.  The contents will be moved to other pull requests that will be more compatible with the head of the development.

**Pull request closed without merging**